FirebaseInstallations/Source/Library/IIDMigration/FIRInstallationsIIDStore.m

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FIRInstallationsIIDStore.h"
+
+#if __has_include(<FBLPromises/FBLPromises.h>)
+#import <FBLPromises/FBLPromises.h>
+#else
+#import "FBLPromises.h"
+#endif
+
+#import <CommonCrypto/CommonDigest.h>
+#import "FIRInstallationsErrorUtil.h"
+
+static NSString *const kFIRInstallationsIIDKeyPairPublicTagPrefix =
+    @"com.google.iid.keypair.public-";
+static NSString *const kFIRInstallationsIIDKeyPairPrivateTagPrefix =
+    @"com.google.iid.keypair.private-";
+static NSString *const kFIRInstallationsIIDCreationTimePlistKey = @"|S|cre";
+
+@implementation FIRInstallationsIIDStore
+
+- (FBLPromise<NSString *> *)existingIID {
+  return [FBLPromise onQueue:dispatch_get_global_queue(QOS_CLASS_UTILITY, 0)
+                          do:^id _Nullable {
+                            if (![self hasPlistIIDFlag]) {
+                              return nil;
+                            }
+
+                            NSData *IIDPublicKeyData = [self IIDPublicKeyData];
+                            return [self IIDWithPublicKeyData:IIDPublicKeyData];
+                          }]
+      .validate(^BOOL(NSString *_Nullable IID) {
+        return IID.length > 0;
+      });
+}
+
+- (FBLPromise<NSNull *> *)deleteExistingIID {
+  return [FBLPromise onQueue:dispatch_get_global_queue(QOS_CLASS_UTILITY, 0)
+                          do:^id _Nullable {
+                            NSError *error;
+                            if (![self deleteIIDFlagFromPlist:&error]) {
+                              return error;
+                            }
+
+                            if (![self deleteIID:&error]) {
+                              return error;
+                            }
+
+                            return [NSNull null];
+                          }];
+}
+
+#pragma mark - IID decoding
+
+- (NSString *)IIDWithPublicKeyData:(NSData *)publicKeyData {
```

Comment:
should we expose this method in iid as an internal public API so fis can call directly?
Or the goal is for FIS not to depend on iid and figure it out inside the keychain directly?

Reply:
Yeah, the idea is that IID depends on FIS, but FIS does not have any code dependencies on IID. Because of that we have to port part of the IID logic to be able to read Instance ID stored by IID SDK.